### PR TITLE
Add transition guard to state()

### PR DESCRIPTION
**Pull request by slmjkdbtl**
As proposed in [this comment](https://github.com/replit/kaboom/pull/346#pullrequestreview-789795171)

```js
const s = state("move", [ "idle", "attack", "move", ], {
    "idle": [ "move", "attack" ],
    "attack": "idle",
    "move": "attack",
})

// define event for specific transitions
s.onStateTransition("move", "attack", throwFeces)
```

### CODE REVIEW

File: src/kaboom.ts [hunk 1]
```diff
@@ -2159,16 +2163,34 @@ function state(initState: string, stateList?: string[]): StateComp {
...
+			const oldState = this.state;
+			checkTransition(oldState, state);
+			trigger("leave", oldState, ...args);
 			this.state = state;
-			trigger("enter", this.state, ...args);
+			trigger("enter", state, ...args);
+			trigger("enter", `${oldState} -> ${state}`, ...args);
+		},
+		onStateTransition(from: string, to: string, action: () => void) {
+			on("enter", `${from} -> ${to}`, action);
```
Comment (slmjkdbtl) [inline src/kaboom.ts:2193]: a bit hacky, but i kinda like it 🤔

File: src/kaboom.ts [hunk 2]
```diff
@@ -2159,16 +2163,34 @@ function state(initState: string, stateList?: string[]): StateComp {
...
+	function checkTransition(from: string, to: string) {
+		if (!transitions?.[from]) {
+			return
+		}
+		const available = typeof transitions[from] === "string"
+			? [transitions[from]]
+			: transitions[from] as string[];
+		if (!available.includes(to)) {
+			throw new Error(`Cannot transition state from "${from}" to "${state}". Available transitions: ${available.map((s) => `"${s}"`).join(", ")}`);
+		}
```
Comment (masad-frost) [inline src/kaboom.ts:2175]: I wonder if it's more friendly if we return a boolean indicating success or failure.

File: src/kaboom.ts (hunk 1 above)
Comment (masad-frost) [reply src/kaboom.ts:2193]: Haha

File: src/kaboom.ts [hunk 3]
```diff
@@ -2159,16 +2163,34 @@ function state(initState: string, stateList?: string[]): StateComp {
...
 	}
 
+	function checkTransition(from: string, to: string) {
+		if (!transitions?.[from]) {
+			return
+		}
+		const available = typeof transitions[from] === "string"
+			? [transitions[from]]
+			: transitions[from] as string[];
+		if (!available.includes(to)) {
```
Comment (masad-frost) [inline src/kaboom.ts:2173]: Seems unnecessary to create an array here.

```suggestion
		const isAvailable = typeof transitions[from] === "string"
			? transitions[from] === to
			: transitions[from].includes(to)
		if (!isAvailable) {
```

**Review by masad-frost — APPROVED** [review]
:pixel_heart:

File: src/kaboom.ts (hunk 3 above)
Comment (slmjkdbtl) [reply src/kaboom.ts:2173]: Absolutely

Comment (slmjkdbtl) [reply src/kaboom.ts:2173]: Oh the array is used [later](https://github.com/replit/kaboom/blob/b0d1f50d68fb0f90d72bc325df5c480376bb222a/src/kaboom.ts#L2174) to print the available transition destinations

Comment (masad-frost) [reply src/kaboom.ts:2173]: Gotcha

**Pull request merged**